runtime/sema/type_tags.go

```diff
@@ -376,7 +376,16 @@ func LeastCommonSuperType(types ...Type) Type {
 		join = join.Or(typ.Tag())
 	}
 
-	return findCommonSuperType(join, types...)
+	supertype := findCommonSuperType(join, types...)
+
+	// 'Any' is a valid type, since it's the supertype of all types.
+	// However, in the context of checker, 'Any' is not a possible
+	// type for expressions. Hence, return 'InvalidType'.
+	if supertype == AnyType {
+		return InvalidType
+	}
```

Comment:
👍 

Maybe flag this as being important, and also describe why `Any` is not a valid type:

```suggestion
	// NOTE: Important:
	// 'Any' is a valid inferred type, since it's the supertype of all types.
	// However, in the context of checker, 'Any' is not a possible type for expressions,
	// as it mixes value-kinded and resource-kinded types, leading to undefined behaviour.
	// Hence, return 'InvalidType'.
	if supertype == AnyType {
		return InvalidType
	}
```